Lista de dicas/técnicas:
1. Algoritmo de Machine Learning - KNN Regression
2. Divisão do dataset train / test
3. Balaceamento do dataset
4. Tranformação dos dados
5. Modelo (Baseline)
6. Métricas de performance para modelos de Classificação

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
df = pd.read_csv('df_model.csv')

In [3]:
df.head()

Age  BusinessTravel  DistanceFromHome  EnvironmentSatisfaction  \
0   25               1                 1                        1   
1   44               1                 1                        2   
2   27               1                17                        4   
3   40               2                 5                        4   
4   24               1                24                        4   

   JobInvolvement  JobLevel  JobSatisfaction  MonthlyIncome  OverTime  \
0               3         2                3           4898         0   
1               2         3                1          10482         0   
2               3         1                3           2394         1   
3               2         3                1          10475         1   
4               3         1                2           3597         0   

   RelationshipSatisfaction  ...  JobRole_Laboratory Technician  \
0                         4  ...                              0   
1                         4  ...                              0   
2                         4  ...                              0   
3                         3  ...                              0   
4                         4  ...                              1   

   JobRole_Manager  JobRole_Manufacturing Director  JobRole_Research Director  \
0                0                               1                          0   
1                0                               0                          0   
2                0                               0                          0   
3                0                               0                          0   
4                0                               0                          0   

   JobRole_Research Scientist  JobRole_Sales Executive  \
0                           0                        0   
1                           0                        0   
2                           1                        0   
3                           0                        1   
4                           0                        0   

   JobRole_Sales Representative  MaritalStatus_Divorced  \
0                             0                       0   
1                             0                       0   
2                             0                       0   
3                             0                       0   
4                             0                       0   

   MaritalStatus_Married  MaritalStatus_Single  
0                      1                     0  
1                      1                     0  
2                      0                     1  
3                      1                     0  
4                      0                     1  

[5 rows x 36 columns]

In [4]:
df.tail()

Age  BusinessTravel  DistanceFromHome  EnvironmentSatisfaction  \
1171   34               0                 4                        3   
1172   19               1                 3                        2   
1173   19               1                 9                        3   
1174   31               2                 3                        2   
1175   23               1                12                        4   

      JobInvolvement  JobLevel  JobSatisfaction  MonthlyIncome  OverTime  \
1171               3         2                3           6538         0   
1172               3         1                2           1483         0   
1173               3         1                1           2552         0   
1174               3         3                3          13675         0   
1175               3         1                4           2647         0   

      RelationshipSatisfaction  ...  JobRole_Laboratory Technician  \
1171                         1  ...                              0   
1172                         4  ...                              1   
1173                         3  ...                              0   
1174                         1  ...                              0   
1175                         3  ...                              1   

      JobRole_Manager  JobRole_Manufacturing Director  \
1171                0                               0   
1172                0                               0   
1173                0                               0   
1174                0                               0   
1175                0                               0   

      JobRole_Research Director  JobRole_Research Scientist  \
1171                          0                           0   
1172                          0                           0   
1173                          0                           1   
1174                          1                           0   
1175                          0                           0   

      JobRole_Sales Executive  JobRole_Sales Representative  \
1171                        1                             0   
1172                        0                             0   
1173                        0                             0   
1174                        0                             0   
1175                        0                             0   

      MaritalStatus_Divorced  MaritalStatus_Married  MaritalStatus_Single  
1171                       0                      1                     0  
1172                       0                      0                     1  
1173                       0                      0                     1  
1174                       1                      0                     0  
1175                       0                      0                     1  

[5 rows x 36 columns]

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1176 entries, 0 to 1175
Data columns (total 36 columns):
 #   Column                             Non-Null Count  Dtype
---  ------                             --------------  -----
 0   Age                                1176 non-null   int64
 1   BusinessTravel                     1176 non-null   int64
 2   DistanceFromHome                   1176 non-null   int64
 3   EnvironmentSatisfaction            1176 non-null   int64
 4   JobInvolvement                     1176 non-null   int64
 5   JobLevel                           1176 non-null   int64
 6   JobSatisfaction                    1176 non-null   int64
 7   MonthlyIncome                      1176 non-null   int64
 8   OverTime                           1176 non-null   int64
 9   RelationshipSatisfaction           1176 non-null   int64
 10  StockOptionLevel                   1176 non-null   int64
 11  TotalWorkingYears                  1176 non-null   int64
 12  TrainingTimesLastYea

In [6]:
df.shape

(1176, 36)

In [7]:
df.describe()

Age  BusinessTravel  DistanceFromHome  EnvironmentSatisfaction  \
count  1176.000000     1176.000000       1176.000000              1176.000000   
mean     36.964286        1.083333          9.097789                 2.728741   
std       9.148817        0.537792          8.002540                 1.096404   
min      18.000000        0.000000          1.000000                 1.000000   
25%      30.000000        1.000000          2.000000                 2.000000   
50%      36.000000        1.000000          7.000000                 3.000000   
75%      43.000000        1.000000         14.000000                 4.000000   
max      60.000000        2.000000         29.000000                 4.000000   

       JobInvolvement     JobLevel  JobSatisfaction  MonthlyIncome  \
count     1176.000000  1176.000000      1176.000000    1176.000000   
mean         2.725340     2.071429         2.739796    6562.642007   
std          0.708781     1.118510         1.109486    4780.847303   
min          1.000000     1.000000         1.000000    1009.000000   
25%          2.000000     1.000000         2.000000    2870.000000   
50%          3.000000     2.000000         3.000000    4968.000000   
75%          3.000000     3.000000         4.000000    8500.000000   
max          4.000000     5.000000         4.000000   19973.000000   

          OverTime  RelationshipSatisfaction  ...  \
count  1176.000000               1176.000000  ...   
mean      0.274660                  2.702381  ...   
std       0.446533                  1.086017  ...   
min       0.000000                  1.000000  ...   
25%       0.000000                  2.000000  ...   
50%       0.000000                  3.000000  ...   
75%       1.000000                  4.000000  ...   
max       1.000000                  4.000000  ...   

       JobRole_Laboratory Technician  JobRole_Manager  \
count                    1176.000000      1176.000000   
mean                        0.178571         0.070578   
std                         0.383156         0.256228   
min                         0.000000         0.000000   
25%                         0.000000         0.000000   
50%                         0.000000         0.000000   
75%                         0.000000         0.000000   
max                         1.000000         1.000000   

       JobRole_Manufacturing Director  JobRole_Research Director  \
count                     1176.000000                1176.000000   
mean                         0.102041                   0.056122   
std                          0.302831                   0.230256   
min                          0.000000                   0.000000   
25%                          0.000000                   0.000000   
50%                          0.000000                   0.000000   
75%                          0.000000                   0.000000   
max                          1.000000                   1.000000   

       JobRole_Research Scientist  JobRole_Sales Executive  \
count                 1176.000000              1176.000000   
mean                     0.189626                 0.223639   
std                      0.392172                 0.416860   
min                      0.000000                 0.000000   
25%                      0.000000                 0.000000   
50%                      0.000000                 0.000000   
75%                      0.000000                 0.000000   
max                      1.000000                 1.000000   

       JobRole_Sales Representative  MaritalStatus_Divorced  \
count                   1176.000000             1176.000000   
mean                       0.055272                0.221939   
std                        0.228608                0.415727   
min                        0.000000                0.000000   
25%                        0.000000                0.000000   
50%                        0.000000                0.000000   
75%                        0.000000              

Podemos ver que existem features que vão de 18 a 60, como 'Age', e outras que vão de 1009 a 19973, como 'MonthlyIncome'

## **2. Divisão do dataset train / test**

Comentários

Cuidado na hora de separar o X, pode conter algum id (identificador). Ele é nocivo ao modelo pois individualiza as observações, aumentado o risco de overfit. Certifique-se que não há nenhum feature com esta característica.

In [8]:
X = df.drop(columns = 'Attrition')

In [9]:
y = df['Attrition']

In [10]:
X.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1176 entries, 0 to 1175
Data columns (total 35 columns):
 #   Column                             Non-Null Count  Dtype
---  ------                             --------------  -----
 0   Age                                1176 non-null   int64
 1   BusinessTravel                     1176 non-null   int64
 2   DistanceFromHome                   1176 non-null   int64
 3   EnvironmentSatisfaction            1176 non-null   int64
 4   JobInvolvement                     1176 non-null   int64
 5   JobLevel                           1176 non-null   int64
 6   JobSatisfaction                    1176 non-null   int64
 7   MonthlyIncome                      1176 non-null   int64
 8   OverTime                           1176 non-null   int64
 9   RelationshipSatisfaction           1176 non-null   int64
 10  StockOptionLevel                   1176 non-null   int64
 11  TotalWorkingYears                  1176 non-null   int64
 12  TrainingTimesLastYea

In [11]:
y.info()

<class 'pandas.core.series.Series'>
RangeIndex: 1176 entries, 0 to 1175
Series name: Attrition
Non-Null Count  Dtype
--------------  -----
1176 non-null   int64
dtypes: int64(1)
memory usage: 9.3 KB


In [12]:
from sklearn.model_selection import train_test_split

Dois pontos importantes na separação:
 - random_state: Parâmetro que controla a mistura dos dados antes da aplicação da separação. Importante manter o valor deste parâmetro para que seja possível replicar o mesmo dataset, caso necessário;
 - stratify: parâmetro que possibilita que os dados separados tenham a mesma proporção dos dados originais. 

In [13]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)

## **3. Balaceamento do dataset**

Para dados discretos, é importante que os dados de treino estejam balanceados, para evitar viés no modelo. O indicado é uma diferença de 10%.

**3.1 Verificando balanceamento dos dados (variável alvo: 'Attrition')**

In [14]:
y_train.value_counts(normalize=True)

0    0.838298
1    0.161702
Name: Attrition, dtype: float64

Podemos verificar que os dados estão desbalanceados.

Como não aprendemos, neste módulo, técnicas de balanceamento, manteremos os dados desta forma.

## **4. Transformação dos dados**

Para algoritimos baseados em distância, como o KNN, é necessário transformar os dados para que as diferenças de escala não interfiram na qualidade do modelo. Estudamos dois métodos:
 - Padronização (Standardization): Calcula parâmetros estatísticos e converte os dados em realação a quantidade de desvios padrões ele se posiciona em relação à média.
 - Mínimo/Máximo: Defini o valor máximo e mínimo  e converte os dados baseado nesses valores.
 
 
 **- Escolha:** Método de padronização (StandardScaler)
 
 **- Justificativa:** O método Min/Max possui um risco. Após definir escala para os dados de treino, há risco dos dados de teste estarem fora desta escala. Esse risco não existe para o método de padronização.
 
  Além disso, pelo fato da padronização converter os dados em desvios padrões, a interpretação é mais fácil, consequentemente, a detecção de anormalidades.
  
  **- Ponto de atenção:** A tranformação deve ser feita após a divisão do dataset em treino e teste. O motivo é o risco dos dados de treino influenciarem os dados de teste, afinal, os scaler calcula os parâmetros e transforma os dados a partir do dataset original (base). Esse tipo de erro é chamado de DATA LEAK.

In [15]:
from sklearn.preprocessing import StandardScaler

In [16]:
scaler_train = StandardScaler()

In [17]:
X_train_std = scaler_train.fit_transform(X_train)

X_train_std

array([[ 1.30073014,  1.71012312, -0.88237556, ..., -0.54456563,
         1.11509495, -0.69473005],
       [ 1.52111341,  1.71012312, -0.88237556, ..., -0.54456563,
         1.11509495, -0.69473005],
       [ 2.07207156,  1.71012312, -0.00254056, ..., -0.54456563,
         1.11509495, -0.69473005],
       ...,
       [-0.90310247, -0.12703772, -0.75668485, ..., -0.54456563,
        -0.89678462,  1.43940801],
       [-0.90310247, -0.12703772,  0.5002223 , ..., -0.54456563,
         1.11509495, -0.69473005],
       [ 0.19881384,  1.71012312, -0.12823127, ..., -0.54456563,
         1.11509495, -0.69473005]])

In [18]:
X_test_std = scaler_train.fit_transform(X_test)

X_test_std

array([[-0.85910924, -0.27643388,  0.93040117, ...,  2.03234711,
        -0.98319208, -0.6758625 ],
       [ 1.27734788, -0.27643388, -0.29490394, ...,  2.03234711,
        -0.98319208, -0.6758625 ],
       [-1.71369208, -0.27643388,  1.42052322, ..., -0.49204193,
        -0.98319208,  1.47959089],
       ...,
       [ 2.13193072, -2.25335496, -1.03008701, ...,  2.03234711,
        -0.98319208, -0.6758625 ],
       [-0.96593209, -0.27643388, -0.04984292, ..., -0.49204193,
        -0.98319208,  1.47959089],
       [ 1.7046393 , -0.27643388, -0.9075565 , ..., -0.49204193,
        -0.98319208,  1.47959089]])

## **5. Modelo**

In [19]:
from sklearn.neighbors import KNeighborsClassifier

In [20]:
modelo = KNeighborsClassifier(n_neighbors=5)

In [21]:
modelo.fit(X_train_std, y_train)

KNeighborsClassifier()

In [22]:
y_pred = modelo.predict(X_test_std)
y_pred

C:\Users\decot\anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdims` to True or False to avoid this warning.
  mode, _ = stats.mode(_y[neigh_ind, k], axis=1)


array([0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0], dtype=int64)

## 6. Métricas de performance

In [23]:
from sklearn.metrics import classification_report

In [24]:
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.84      0.96      0.90       198
           1       0.20      0.05      0.08        38

    accuracy                           0.81       236
   macro avg       0.52      0.51      0.49       236
weighted avg       0.74      0.81      0.77       236



Acurácia de 80%, porém não siginifica nada para o nosso modelo, visto que poucas pessoas deixam a empresam em relação ao total.
Similar ao exemplo do modelo de contratação de candidato, dado em aula pelo professor Kuma.

Exemplo: Se o modelo tem que prever qual, dentre os 100 candidatos, será contratado para a única vaga disponível, e ele errar o candidato, a acurácia será de 99%, já que ele acertou a predição para 99 dos 100 candidatos.

Acurácia é mais signifcativa em dados balanceados. Nesse caso, há mais segurança de que o modelo acertou as classificações. 

**6.1**- Matriz de confusão
<img src=https://diegonogare.net/wp-content/uploads/2020/04/matrizConfusao-600x381.png height="400" width="400">

Note que a diagonal principal são as observações que o modelo acertou! Temos:

- Verdadeiros Positivos (VP): classificação correta da classe positivo;
- Verdadeiros Negativos (VN): classificação correta da classe negativo;
- Falsos Positivos (FP, erro tipo I): correto: negativo. Previsto: positivo.
- Falsos Negativos (FN, erro tipo II): correto: positivo. Previsto: negativo.

Em nosso case, as classificações são as seguintes:

    0 - Negativo - Permaneceu na empresa (retenção);
    1 - Positivo - Saiu da empresa (turnover).

Como nosso objetivo principal é reduzir o turnover, queremos evitar, prioritariamente, o **Falso Negativo (FN)**. Ou seja, quando nossa previsão foi que ele não sairia, mas ele sai. Erro tipo II.